src/util/bson.hpp

```diff
@@ -0,0 +1,1218 @@
+        return true;
+    }
+
+    /*!
+    @brief an unsigned integer number was read
+    @param[in] val  unsigned integer value
+    @return whether parsing should proceed
+    */
+    bool number_unsigned(number_unsigned_t val) override {
+        auto instruction = m_instructions.top();
+        m_instructions.pop();
+        switch (instruction.type) {
+            case state_t::max_key:
+                m_marks.top().push_back({instruction.key, max_key});
+                m_instructions.push({state_t::skip});
+                break;
+            case state_t::min_key:
+                m_marks.top().push_back({instruction.key, min_key});
+                m_instructions.push({state_t::skip});
+                break;
+            case state_t::timestamp_i:
+                if (m_marks.top().back().first == instruction.key) {
+                    auto ts = std::get<Timestamp>(m_marks.top().back().second);
+                    m_marks.top().pop_back();
+                    m_marks.top().push_back({instruction.key, Timestamp(ts.get_seconds(), 1)});
+
+                    // pop vestigal timestamp instruction
+                    m_instructions.pop();
+                    m_instructions.push({state_t::skip});
+                    m_instructions.push({state_t::skip});
+                } else {
+                    m_marks.top().push_back({instruction.key, Timestamp(0, 1)});
+                    instruction.type = state_t::timestamp;
+                }
+                break;
+            case state_t::timestamp_t:
+                if (m_marks.top().back().first == instruction.key) {
+                    auto ts = std::get<Timestamp>(m_marks.top().back().second);
+                    m_marks.top().pop_back();
+                    m_marks.top().push_back({instruction.key, Timestamp(val, ts.get_nanoseconds())});
+
+                    // pop vestigal teimstamp instruction
+                    m_instructions.pop();
+                    m_instructions.push({state_t::skip});
+                    m_instructions.push({state_t::skip});
+                } else {
+                    m_marks.top().push_back({instruction.key, Timestamp(val, 0)});
+                    instruction.type = state_t::timestamp;
+                }
+                break;
+            default:
+                throw BsonError(util::format("Invalid state: %1 for JSON instruction number_unsigned: %2",
+                                             state_to_string(instruction.type),
+                                             val));
+        }
+        return true;
+    }
+
+    /*!
+    @brief an floating-point number was read
+    @param[in] val  floating-point value
+    @param[in] s    raw token value
+    @return whether parsing should proceed
+    */
+    bool number_float(number_float_t, const string_t&) override {
+        throw BsonError(util::format("Invalid SAX instruction for canonical extended json: 'number_float'"));
+        return true;
```

Comment:
I think this will cause a warning for unreachable code because we always throw (there's other place like this too)